crates/hyperdrive-math/src/lp/remove.rs

```diff
@@ -1,6 +1,312 @@
+use ethers::types::{I256, U256};
+use eyre::{eyre, Result};
+use fixed_point::{fixed, FixedPoint};
+
+use super::math::SHARE_PROCEEDS_MAX_ITERATIONS;
 use crate::State;
 
-impl State {}
+impl State {
+    /// Allows an LP to burn shares and withdraw from the pool.
+    pub fn calculate_remove_liquidity(
+        &self,
+        current_block_timestamp: U256,
+        active_lp_total_supply: FixedPoint,
+        withdrawal_shares_total_supply: FixedPoint,
+        lp_shares: FixedPoint,
+        min_output_per_share: FixedPoint,
+        minimum_transaction_amount: FixedPoint,
+    ) -> Result<(FixedPoint, FixedPoint)> {
+        // Ensure that the amount of LP shares to remove is greater than or
+        // equal to the minimum transaction amount.
+        if lp_shares < minimum_transaction_amount {
+            return Err(eyre!("Minimum transaction amount not met"));
+        }
+
+        // Get vault share price and proceed without checkpoint.
+        let vault_share_price = self.vault_share_price();
+
+        let mut state = self.clone();
+        // Burn the LP's shares.
+        state.info.lp_total_supply -= lp_shares.into();
+        let active_lp_total_supply = active_lp_total_supply - lp_shares;
+
+        // Mint an equivalent amount of withdrawal shares.
+        let withdrawal_shares_total_supply = withdrawal_shares_total_supply + lp_shares;
+
+        // Redeem as many of the withdrawal shares as possible.
+        let (proceeds, withdrawal_shares_redeemed) = state.calculate_redeem_withdrawal_shares(
+            current_block_timestamp,
+            active_lp_total_supply,
+            withdrawal_shares_total_supply,
+            lp_shares,
+            vault_share_price,
+            min_output_per_share,
+        )?;
+        let withdrawal_shares = lp_shares - withdrawal_shares_redeemed;
+
+        Ok((proceeds, withdrawal_shares))
```

Comment:
Elsewhere (e.g. [short/open](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/short/open.rs#L160-L177) and [long/open](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/long/open.rs#L53-L76)) we have been separating functions that compute the new state and new deltas. This case is a little tricky... You permute state as part of the process (adjusting lp total supply), which I assume is then considered when calculating redeem withdrawal shares.

Do you think it makes sense to include the new `state` in the return tuple, or you could return a `lp_supply_delta` variable to be used in a later `calculate_pool_state_after_remove_liquidity` function. Or don't add an extra variable and re-compute it in a dedicated `calculate_pool_deltas*` fn.

Thoughts?

Reply:
If there's a need for it we can, until then I'd say leave as is?

Reply:
Ok I actually added this in because it helped with some extra test scenarios.